contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -4290,3 +4317,284 @@ sp_reset_connection_internal(PG_FUNCTION_ARGS)
+	 * full-length header.
+	 */
+	oldcontext = MemoryContextSwitchTo(XMLHashtabContext);
+	new_xml_data = (xmltype *) palloc(VARSIZE_ANY_EXHDR(xml_data) + VARHDRSZ);
+	MemoryContextSwitchTo(oldcontext);
+	// new_xml_data = (xmltype *) palloc(VARSIZE_ANY_EXHDR(xml_data) + VARHDRSZ);
+
+	SET_VARSIZE(new_xml_data, VARSIZE_ANY_EXHDR(xml_data) + VARHDRSZ);
+
+	/*
+	 * VARDATA is a pointer to the data region of the new struct.  The source
+	 * could be a short datum, so retrieve its data through VARDATA_ANY.
+	 */
+	memcpy(VARDATA(new_xml_data),		/* destination */
+		   VARDATA_ANY(xml_data),		/* source */
+		   VARSIZE_ANY_EXHDR(xml_data));	/* how many bytes */
+
+	return new_xml_data;	   
+}
+
+int 
+pltsql_insert_xml_handle_entry(xmltype *xml_data, xmltype *ns_data, int xml_data_length, int ns_data_length)
+{
+    XMLHandleHashEnt *hentry;
+    bool found;
+    int  handle;
+    int  document_id;
+    
+    handle = get_next_xml_handle();      /* get the next handle */
+    document_id = 2*handle - 1;          /* document_id is always odd and unique */
+   
+    hentry = (XMLHandleHashEnt *) hash_search(XMLHandleHashTable, &document_id, 
+                                             HASH_ENTER, &found);
+    if (found)
+        ereport(ERROR, errmsg("Duplicate XML handles"));
+
+    hentry->xml_data = copyXMLType(xml_data);
+    hentry->ns_data = copyXMLType(ns_data);
+    hentry->document_id = document_id;
+   	hentry->original_document_size_bytes = xml_data_length;  /* store the size of the xml text and xpath namespaces */
+    hentry->original_namespace_document_size_bytes = ns_data_length;
+    
+    if (ns_data_length > 0)
+    {
+        hentry->namespace_id = 2*handle;    /* namespace_id is always even and unique if namespace is present */
+        hentry->is_namespace_null = false;
+    }
+    else
+    {
+        hentry->namespace_id = 0;  
+        hentry->is_namespace_null = true;
+    }
+
+    return hentry->document_id;
+}
+
+void 
+pltsql_delete_xml_handle_entry(int document_id)
+{
+    XMLHandleHashEnt *hentry;
+
+    /* First find the entry */
+    hentry = (XMLHandleHashEnt *) hash_search(XMLHandleHashTable, &document_id, 
+                                             HASH_FIND, NULL);
+    if (hentry == NULL)
+    {
+        ereport(ERROR, errmsg("Could not find prepared statement with handle %d", document_id));
+    }
+
+    /* Remove the entry from hash table */
+    hash_search(XMLHandleHashTable, &document_id, HASH_REMOVE, NULL);
+}
+
+/* Function to initialise the hash table */
+Datum
+init_tsql_xml_handle_hash_tab(PG_FUNCTION_ARGS)
+{
+    /* Skip to set up if already created */
+    if (XMLHandleHashTable != NULL)
+        PG_RETURN_INT32(0);
+
+    pltsql_create_xml_handle_htab();
+    
+    PG_RETURN_INT32(0);
+}
+
+/*
+ * reset_cached_xml_handle:
+ *		Cleans up all the stale states and resets the xml handles.
+ *		This function should be called when a connection is cancelled or terminated.
```

Comment:
Just say connection is reset or connection is terminated. There is no thing like "connection is cancelled"